src/connectionbenchmark/benchmarktaskdownload.cpp

```diff
@@ -31,43 +40,80 @@ void BenchmarkTaskDownload::handleState(BenchmarkTask::State state) {
   logger.debug() << "Handle state" << state;
 
   if (state == BenchmarkTask::StateActive) {
-    m_request = NetworkRequest::createForGetUrl(this, m_fileUrl);
-    connect(m_request, &NetworkRequest::requestUpdated, this,
+    // Start DNS resolution
+#if !defined(MVPN_DUMMY)
+    m_dnsLookup.setNameserver(QHostAddress(MULLVAD_DEFAULT_DNS));
+#endif
+    m_dnsLookup.lookup();
+  } else if (state == BenchmarkTask::StateInactive) {
+    for (NetworkRequest* request : m_requests) {
+      request->abort();
+    }
+    m_requests.clear();
+  }
+}
+
+void BenchmarkTaskDownload::dnsLookupFinished() {
+  if (m_dnsLookup.error() != QDnsLookup::NoError) {
+    logger.error() << "DNS Lookup Failed:" << m_dnsLookup.errorString();
+    emit finished(0, true);
+    emit completed();
+    return;
+  }
+  if (m_dnsLookup.hostAddressRecords().isEmpty()) {
+    logger.error() << "DNS Lookup Failed: no records";
+    emit finished(0, true);
+    emit completed();
+    return;
+  }
+  logger.debug() << "DNS Lookup Finished";
+
+  for (const QDnsHostAddressRecord& record : m_dnsLookup.hostAddressRecords()) {
```

Comment:
What about if we have been cancelled in the meantime?

Reply:
Ah yes, I neglected to think of that possibility. I'll add another commit to address it.